examples/shell/shell_common/cmd_ping.cpp

```diff
@@ -129,12 +129,60 @@ class PingArguments
 } gPingArguments;
 
 Protocols::Echo::EchoClient gEchoClient;
+Transport::AdminPairingTable gAdmins;
 
-bool EchoIntervalExpired(void)
+CHIP_ERROR SendEchoRequest(streamer_t * stream);
+
+Transport::PeerAddress GetEchoPeerAddress()
 {
-    uint64_t now = System::Timer::GetCurrentEpoch();
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    if (gPingArguments.IsUsingTCP())
+    {
+        return Transport::PeerAddress::TCP(gDestAddr, gPingArguments.GetEchoPort());
+    }
+    else
+#endif
+    {
 
-    return (now >= gPingArguments.GetLastEchoTime() + gPingArguments.GetEchoInterval());
+        return Transport::PeerAddress::UDP(gDestAddr, gPingArguments.GetEchoPort(), INET_NULL_INTERFACEID);
+    }
+}
+
+void Shutdown()
+{
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    if (gPingArguments.IsUsingTCP())
+    {
+        gTCPManager.Disconnect(GetEchoPeerAddress());
+    }
+    gTCPManager.Close();
+#endif
+    gUDPManager.Close();
+
+    gEchoClient.Shutdown();
+    gExchangeManager.Shutdown();
+    gSessionManager.Shutdown();
+}
+
+void EchoTimerHandler(chip::System::Layer * systemLayer, void * appState, chip::System::Error error)
+{
+    if (gPingArguments.GetEchoRespCount() != gPingArguments.GetEchoCount())
+    {
+        streamer_printf(streamer_get(), "No response received\n");
+        gPingArguments.SetEchoRespCount(gPingArguments.GetEchoCount());
+    }
+    if (gPingArguments.GetEchoCount() < gPingArguments.GetMaxEchoCount())
+    {
+        CHIP_ERROR err = SendEchoRequest(streamer_get());
+        if (err != CHIP_NO_ERROR)
+        {
+            streamer_printf(streamer_get(), "Send request failed: %s\n", ErrorStr(err));
```

Comment:
Shutdown will not get called if we fall under this condition, and there might be also a pending timer

Reply:
How this issue get resolved? Suppose the first SendEchoRequest succeed, and the second SendEchoRequest failed with error within EchoTimerHandler.  How Shutdown() is triggered in this case